physlr/physlr.py

```diff
@@ -1021,107 +1024,190 @@ def assign_read_molecule(mxs, moltomxs, mol_counts, bx):
+            sub_graph_copy.remove_edges_from(edges_to_remove_barcode)
+            cos_components = list(nx.connected_components(sub_graph_copy))
+            cos_components.sort(key=len, reverse=True)
+            for com in cos_components:
+                # if len(com) > 1:
+                communities.append(com)
+        return communities
+
+    @staticmethod
+    def community_detection_maximum_spanning_tree(g, node_set):
+        """Find MST. Collect clusters of nodes around a hub.
+        Merge if appropriate. Return communities."""
+        sub_mst = nx.maximum_spanning_tree(g.subgraph(node_set), weight="n")
+        hubs = [i[0] for i in sub_mst.degree() if i[1] > 2]
+        communities = \
+            [set(sub_mst.subgraph(
+                set(sub_mst.neighbors(hub)).union({hub})).nodes())
+             for hub in hubs]
+        return Physlr.merge_communities(g, communities)
+
+    @staticmethod
+    def merge_communities(g, communities, node_set=0, strategy=0, cutoff=8):
+        """Merge communities if appropriate."""
+        mode = 1
+        if len(communities) == 1 and (node_set == 0 or strategy == 1):
+            return communities
+        if strategy == 1:  # Merge by Initializing Louvain with the communities
+            return Physlr.community_detection_louvain(g, node_set, communities)
+        # Ad-hoc Merge (default - strategy = 0)
+        merge_network = nx.Graph()
+        for i in range(len(communities)):
+            merge_network.add_node(i)
+        for i, com1 in enumerate(communities):
+            for k, com2 in enumerate(communities):
+                if i < k:
+                    if mode == 1:  # disjoint input communities.
+                        if nx.number_of_edges(
+                                g.subgraph(com1.union(com2))) - \
+                                nx.number_of_edges(g.subgraph(com1)) - \
+                                nx.number_of_edges(g.subgraph(com2)) > cutoff:
+                            merge_network.add_edge(i, k)
+                    else:  # overlapping input communities.
+                        if nx.number_of_edges(
+                                g.subgraph(com1.union(com2))) - \
+                                len(set(g.subgraph(com1).edges()).union(
+                                    set(g.subgraph(com2).edges()))) \
+                                > cutoff:
+                            merge_network.add_edge(i, k)
+        return [{barcode for j in i for barcode in communities[j]}
+                for i in nx.connected_components(merge_network)]
 
     @staticmethod
     def determine_molecules(g, u, strategy):
-        "Assign the neighbours of this vertex to molecules."
-        if strategy == 2:
-            return Physlr.determine_molecules_k_clique_communities(g, u)
-        if strategy == 3:
-            return Physlr.determine_molecules_louvain(g, u)
-        if strategy == 4:
-            return Physlr.determine_molecules_cosine_of_squared(g, u)
```

Comment:
Please add these functions `Physlr.determine_molecules_*`. They can be defined in terms of `Physlr.community_detection_*` if you like. For example:
```python
def determine_molecules_k_clique_communities(g, u):
    return [community
        for bi_connected_component in
        Physlr.community_detection_biconnected_components(g, set(g.neighbors(u)))
        for community in
        Physlr.community_detection_k_clique(g, bi_connected_component)]
end
```